bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ArrayValueImpl.java

```diff
@@ -1206,15 +1210,14 @@ private void shiftArray(int index, Object arr) {
     private void unshiftArray(long index, int unshiftByN, int arrLength) {
         int lastIndex = size() + unshiftByN - 1;
         prepareForConsecutiveMultiAdd(lastIndex, arrLength);
-        Object arr = getArrayFromType(elementType.getTag());
-
         if (index > lastIndex) {
             throw BLangExceptionHelper.getRuntimeException(
                     getModulePrefixedReason(ARRAY_LANG_LIB, INDEX_OUT_OF_RANGE_ERROR_IDENTIFIER),
                     RuntimeErrors.INDEX_NUMBER_TOO_LARGE, index);
         }
-
         int i = (int) index;
+        ensureCapacity(this.size + unshiftByN, this.size);
```

Comment:
Does "if (index > lastIndex)" condition  not handle capacity?

Reply:
no. 
As per the current implementation, we first shift the array by the no. of values and then add the new values to the array. The error happens when we shift the array due to the default array size (=100) we have for open arrays. 
The `if (index > lastIndex)` check only validates the size of the BArray, not the size of internal array (Eg. : `intValues`, `refValues` etc.).